### PR TITLE
WIP: Introduce Overflow Checking for ^(::Integer, ::Integer)

### PR DESCRIPTION
**Pull request by Keno**
I haven't updated the tests yet, mostly because we need to get a consensus on what to do.
The thinking behind this is that by far the most common case of people running into overflows is trying to do things like `10^x`, when they meant `10.0^x`. Since this functionality is very well isolated, we can add checking without generally killing performance. There's a couple of details to work out though:
- What to do about small powers ^2, ^3, etc. @StefanKarpinski expressed a preference for those to keep behaving like `x*x`, `x*x*x`.
- What to do about literals. Do we keep referential transparency (e.g. do the literal ^2, ^3 no check for overflow, but the non-literal one does).
- Are we ok with this? Initial performance numbers look like about a 10-20% performance penalty for this change.

cc @JeffBezanson @StefanKarpinski @stevengj @timholy

### CODE REVIEW

**Comment by stevengj** [comment]
I don't like the idea of slowing down a really primitive operation like `^`, particular when the overflow checking is only done sporadically.

**Comment by Keno** [comment]
Is integer pow really used that often though? The range of usable inputs is actually fairly small (particularly for say Int32).

**Comment by TotalVerb** [comment]
It seems a little inconsistent to me to only check overflow in this function. Wouldn't it be better to have a flag/environment/module where all functions are overflowed check?

**Comment by Keno** [comment]
It would be, but we can't implement that in an efficient manner. This is supposed to catch a case where this happens very frequently that's implementable without terrible performance regressions.

**Comment by timholy** [comment]
If it's only a 10-20% slowdown, and if it seems unlikely that future changes in CPUs will significantly widen the gap, then I'm certainly willing to consider it. I'd think we'd want a consistent policy with regards to `^`, however, even for small integer powers. (Meaning that, in contrast to @StefanKarpinski's suggestion above, I'd say if we have it for any powers of integers we should have it for all.)

File: base/intfuncs.jl [hunk 1]
```diff
@@ -171,18 +172,23 @@ function power_by_squaring(x_, p::Integer)
...
         while (t -= 1) >= 0
-            x *= x
+            x, o′ = mul_with_overflow(x, x)
+            o |= o′
         end
-        y *= x
+        y, o′ = mul_with_overflow(y, x)
+        o |= o′
     end
+    o && throw(OverflowError)
```
Comment (tkelman) [inline base/intfuncs.jl]: probably want an instance, not the type

Comment (Keno) [inline base/intfuncs.jl]: yes, already fixed locally. I just figured I'd start the policy discussion with some code and numbers ;).

**Comment by Keno** [comment]
I'm happy to make it always checked.

**Comment by timholy** [comment]
I wonder if we could use `@fastmath` to skip the check? I'm mostly thinking of `@simd`, where any branch in the inner loop prevents vectorization.

**Comment by PallHaraldsson** [comment]
"where any branch in the inner loop prevents vectorization."

Do you need that? I'm not sure with LLVM, but in assembly can't you get a flag; and can you then OR them together, and check out of the loop?

**Comment by Keno** [comment]
Does anybody have an example of real code where this is performance critical? It would be good to look at some real world examples, because I suspect in a lot of cases we might be able to hoist the check anyway.

**Comment by djsegal** [comment]
Any word on this before it gets absorbed into the abyss?

**Comment by stevengj** [comment]
I'm inclined to think that @Keno is right and that `Int^Int` is rarely performance critical in real code, so that overflow checking is worth the overhead, at least in `^(::IntXX, ::IntXX)`.

**Comment by JeffBezanson** [comment]
I'm not sure this is worth it. It just seems kind of complex/random/undisciplined to check for overflow only in this case. As it is, those who want overflow to be checked at least acknowledge that we made a decision and stuck with it :shrug:

**Comment by oscardssmith** [comment]
The theory here is that powers are the one integer operation that is commonly used and commonly overflows with small-ish inputs. Also, with this implementation, the amount of overhead is approximately 0.

**Comment by perrutquist** [comment]
Maybe this is stating the obvious, but...  

For literal exponents, the overflow check only requires two integer comparisons at runtime, since the bounds on `x` can be pre-computed. For example:

```julia
function literal_pow_overflows(x::Int64, ::Val{p}) where {p}
    p <= 1 && return false
    p >= 64 && return x < -1 || x > 1
    (x < (-3037000499, -2097152, -55108, -6208, -1448, -512, -234, -128, -78, -52, -38, -28, 
        -22, -18, -15, -13, -11, -9, -8, -8, -7, -6, -6, -5, -5, -5, -4, -4, -4, -4, -3, -3, -3, -3, 
        -3, -3, -3, -3, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, 
        -2, -2, -2, -2, -2)[p-1] ||
    x > (3037000499, 2097151, 55108, 6208, 1448, 511, 234, 127, 78, 52, 38, 28, 
         22, 18, 15, 13, 11, 9, 8, 7, 7, 6, 6, 5, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 
         3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 
         2, 2, 2, 1)[p-1])
end
```

(The same method could obviously also be applied to non-literal exponents, although the table lookups would probably be too expensive.)

**Comment by PallHaraldsson** [comment]
I really think we should do something about overflow (since a very common criticism of Julia), even just for power, in some form, and I prefer A:

A.
Simply have a^b return Float64, since in general returning an integer is wrong (and power is the only type-unstable primitive operation), consider the previous design a bug (if b is Unsigned, that's an exception, and can keep status quo).

Then my proposal here (made for physics code, speed-of-light is redundant):
https://github.com/tonyhffong/Lint.jl/issues/271

B.
The code as is has many checks, and it could start with if b > 3 and only overflow check for that.

**Comment by stevengj** [comment]
@PallHaraldsson, my sense is people are broadly receptive to this, but we need someone to step up and write an optimized implementation.

**Comment by perrutquist** [comment]
If this is implemented, then there also needs to be an easy way to opt out from checking when the overflow is intentional. My suggestion would be to let `Base.powermod` accept a type as the modulus, so that `powermod(x,p,typeof(x))` can be used for non-checking `x^p`.

**Comment by LilithHafner** [comment]
I think that the idea to catch common overflows is great, but this is technically breaking because `17^29` is currently a valid way to compute 17 to the power of 29 mod Int64. Given the lack of decisive support for this, I think it is not worth slipping this minor breakage into a 1.x release.

Perhaps this could become part of a larger discussion on overflow behavior come 2.0.

**Comment by oscardssmith** [comment]
triage thinks this is probably too breaking for 1.x

**Comment by vtjnash** [comment]
Just saw this was implemented and merged in a later PR

**Comment by Keno** [comment]
I don't think so:
```
julia> 10^200
0
```

**Comment by oscardssmith** [comment]
It's `Base.checked_pow` in https://github.com/JuliaLang/julia/pull/52849.

**Comment by Keno** [comment]
That's not what this PR does though. This PR turns on overflow checking by default in `^`.

**Comment by vtjnash** [comment]
This PR attempted to do both things at once it seems, which stalled it. We can change the default later, but the majority of the PR looks like it was implemented now

**Comment by PallHaraldsson** [comment]
>this is technically breaking because 17^29 is currently a valid way to compute 17 to the power of 29 mod Int64.

No it isn't, this is mod Int32 on 32-bit computers. So getting two different answers is problematic in itself for me, unless in the case there's actually no truncation. And this is the most dangerous operation.

How about widening always to Int128, to at least have a fast operation (rather than to BigInt...), and then you can mod all you want later...?

**Pull request closed without merging**